multiaddr/multiaddr.py

```diff
@@ -37,9 +39,16 @@ def __init__(self, addr):
             addr : A string-encoded or a byte-encoded Multiaddr
 
         """
-        if isinstance(addr, str):
+        # On Python 2 text string will often be binary anyways so detect the
+        # obvious case of a “binary-encoded” multiaddr starting with a slash
+        # and decode it into text
+        if six.PY2 and isinstance(addr, str) and addr.startswith("/"):
+            import locale
+            addr = addr.decode(locale.getpreferredencoding())
```

Comment:
Could you elaborate more on the reason why we use `locale.getpreferredencoding()` here? The result seems possibly to be different on different platforms.

Reply:
cc @alexander255

Reply:
(I did receive the first message.)

My reasoning for this was basically: Since there is no correct answer, I'll just use the best guess available. For instance, Python 3 [will open all text with this encoding by default](https://docs.python.org/3.7/library/functions.html?highlight=open#open). IMO, it's better than hard-coding ASCII, but you're free to disagree with me on this. (As far as I can tell, all encodings returned by `locale.getpreferredenconding()` are also ASCII-compatible in pratice, but don't quote me on that.)

Reply:
(I'm not familiar with locale a lot, please correct me if I'm wrong)
What I'm worried about is, if some user set their locale manually, `addr.decode(locale.getpreferredencoding())` will decode to a wrong string(or failed to decode), and then be encoded to wrong bytes. If this is a reasonable concern(I'm not so sure whether it is), IMO it is probably better to use 'utf-8' as the encoding, at least enforcing the same encoding in all platforms.

Reply:
You're correct on that concern, I thought about this some more and while I stick to my opinion that “its their problem and *locale.getpreferredencoding()* is considered the standard *solution* for these cases”. You have a point though, and with everything moving to UTF-8 anyways *and* there being other multiformats that already work like this. I think we can reasonably force UTF-8 here. :+1: